### PR TITLE
Add IO overload to Benchmark

### PR DESCRIPTION
**Pull request by j8r**
This change allows benchmarkers to directly send the report to an IO, in order to build a Markdown benchmark report for example.  Before creating a subprocess was the only way to do this.
The `interactive` ivar is no longer needed.

### CODE REVIEW

**Comment by RX14** [comment]
No, the interactive ivar is especially needed now. Outputting VT100 escape sequences unconditionally is bad behaviour, especially when outputting to something like markdown.

**Comment by j8r** [comment]
Or we can simply drop the use of escape sequences

**Comment by j8r** [comment]
We can also drop the use of escape sequences

**Comment by j8r** [comment]
That's right, interactive behavior allows to show the results progressively.

**Comment by j8r** [comment]
I've changed `interactive` to a method overload instead of an ivar, which isn't used anywhere else than in `Benchmal::IPS::Job#report`

File: spec/std/benchmark_spec.cr [hunk 1]
```diff
@@ -12,21 +12,22 @@ describe Benchmark::IPS::Job do
...
+      # the mean should be calculated
+      a.mean.should be > 10
+
+      # one of the reports should be normalized to the fastest but do to the
+      # timer precisison sleep 0.001 may not always be faster than 0.002 so we
+      # don't care which
+      first, second = [a.slower, b.slower].sort
+      first.should eq(1)
+      second.should be > 1
+    end
```
Comment (yxhuvud) [inline spec/std/benchmark_spec.cr]: Shouldn't the old usage still be supported? If so, why change the test instead of adding a new one?

Comment (j8r) [reply spec/std/benchmark_spec.cr]: Because we don't want the output to be sent in the STDOUT, which is the default (as before)

Comment (j8r) [reply spec/std/benchmark_spec.cr]: I can create a new tests

Comment (j8r) [reply spec/std/benchmark_spec.cr]: I can create new tests

**Comment by j8r** [comment]
I've modified the API: if we pass an IO, it must be a valid TTY to send the report interactively in it.
I don't know if it's better to raise of not send the report to the IO.
If we want to generate a custom report, we can use the `report` method.

**Comment by Sija** [comment]
> I've modified the API: if we pass an IO, it must be a valid TTY to send the report interactively in it.

Why? What if you'd like to pipe it to the file?

**Comment by j8r** [comment]
If you pipe to a fine @Sija , it can end up having escape sequences.
Generally if you define a custom IO other the default `STDOUT`, you know what you do and can directly use `Benchmark::IPS` instead of the `Benchmark.ips` wrapper.

**Comment by j8r** [comment]
If you pipe to a file @Sija , it can end up having escape sequences.
Generally if you define a custom IO other the default STDOUT, you know what you do and can directly use Benchmark::IPS instead of the Benchmark.ips wrapper.

**Comment by Sija** [comment]
@j8r `Colorize` module should handle that (escape sequences).

**Comment by j8r** [comment]
All escape sequences aren't about color: this one is used to "delete" lines - (try `printf "\e[2A"`)

**Comment by asterite** [comment]
@j8r Originally this PR was about being able to send the report through an IO, say, a file. But a file is not a tty. And it seems you are raising on non-tty devices. So how does this change support your original idea? I think I might be missing something...

**Comment by Sija** [comment]
> All escape sequences aren't about color: this one is used to "delete" lines - (try `printf "\e[2A"`)

And where is it used?

This PR becomes really messy and your changes are way out of initially defined scope (`class` -> `struct`, tty related changes, etc.).

**Comment by j8r** [comment]
Yes @asterite. I can add an example somewhere in the API docs or in the gitbook. If you want to send the benchmark to a non TTY IO:
```crystal
require "benchmark"

job = Benchmark::IPS::Job.new
job.report "a" do
  sleep 0.2  
end

job.report "b" do
  sleep 0.1
end

job.execute nil

File.open "./report", "w" do |fd|
  job.report  fd
end
```
Content in `./report`:
```
a   4.99  (200.29ms) (± 0.04%)  0.0B/op   2.00× slower
b   9.98  (100.18ms) (± 0.07%)  0.0B/op        fastest
```

**Comment by j8r** [comment]
I've rebased in two commits:
- one to specifically add IO support to Benchmark. This directly relates to TTY checks, etc
- another one add minior optimizations/code improvements here and there (can be another PR of course)

**Comment by asterite** [comment]
This looks weird:

```crystal
job.execute nil
```

Can't we have `execute()` and `execute_interactive(io)`?

**Comment by j8r** [comment]
> And where is it used?

@Sija please read more carefully, see https://github.com/crystal-lang/crystal/blob/master/src/benchmark/ips.cr#L106

`print "\e[#{ran_items.size}A"`

**Comment by j8r** [comment]
@asterite or we can have `execute()` and `execute(io)` with a mandatory IO.

**Comment by asterite** [comment]
Yes, that can work too. But having to pass `nil` is a no-no.

**Comment by Sija** [comment]
@j8r oh, I see. Thanks for pointing me into a right line. Perhaps it should be replaced with `Colorize.reset` (I don't remember if such method actually exists).

**Comment by j8r** [comment]
Not really, I rather think we can have an API for [ANSI escape sequences](https://en.wikipedia.org/wiki/ANSI_escape_code) in general, not only colors - it can control the cursor, the scroll, clear the display etc.
I don't know if it has its place in the stdlib, maybe since Windows officially support them now.
That's beyond this PR, I'll add a topic in the forum.

**Comment by straight-shoota** [comment]
I don't follow why this removes the `interactive: false` behaviour on `Benchmark.ips`. As far as I can see, it should work just fine with a configurable IO (default value `io.tty?`). There seems to be no reason to remove this feature and require to use a lower-level API for this.

**Comment by j8r** [comment]
I have found an other way to generate a benchmark report, PR no longer needed

**Pull request closed without merging**